### PR TITLE
persist: calculate and store stats for all `T: Data` types

### PR DESCRIPTION
**Pull request by danhhz**
This expands the set of column types that persist internally whitelists stats to all types (it was previously only Option<u64> in the skeleton). This might considerably increase the size of the stats metadata, but it's behind a feature flag.

Combined with #18386, we'll get useful stats for a much larger number of ScalarTypes. Finishing off the last few will require a Schema impl to override the stats logic (e.g. for schemaless json stats, understanding Decimals, Lists, etc) is left for followup work. In the meantime, we'll store meaningless stats about their ProtoDatum representations. Actually using these new stats in MFP pushdown is also left for a followup PR.

To accomplish this, we introduce column stats as a first-class concept in mz_persist_types. It's possible this ends up being unnecessary, but it seems like a likely enough bet and we can always pull them back into mz_persist_client if it turns out to be wrong.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by bkirwi — COMMENTED** [review]
A couple nits and one philosophical question, but overall looks great!

File: src/persist-types/src/stats.rs
Comment (danhhz) [reply src/persist-types/src/stats.rs:122]: (assuming I understand the concern correctly) you do get some type safety back in that StructStats only exposes 
 a way to get a column's stats as the associated `Data::Stats` type: `fn col<T: Data>(&self, name: &str) -> Result<Option<&T::Stats>, String>`

I've having trouble imagining what the interface and strongly-typed accessors look like, mind typing up what you're picturing?

File: src/persist-types/src/stats.proto
Comment (bkirwi) [reply src/persist-types/src/stats.proto:24]: > is it to replace ProtoOptionStats option with uint64 null_count?

Yes!

> how does that reduce invalid states?

Currently three interesting states for this field: `ProtoOptionStats` not present, `ProtoOptionStats` present but is  zero / default / unset value, `ProtoOptionStats` present with a nondefault value. The first is invalid for optional stats, and the latter two are invalid for non-optional stats.

The `null_count` idea has two states: field zero / unset or not. Neither is invalid for optional stats, only the latter is invalid for non-optional stats. So 1/4 invalid instead of 3/6!

> also, if that is the suggestion, I don't think it works [...]

I was having trouble thinking why we'd care about the distinction, but I think maybe the idea is that we're expecting to be able to tell whether this is an optional or non-optional stats purely by the proto representation? (vs. needing the schema information in hand to make sense of it.) Which seems reasonable if so.

Comment (danhhz) [reply src/persist-types/src/stats.proto:24]: ah, yeah, at the moment, `<Box<dyn DynStats>>::from_proto` is how we deserialize this and it has to figure out the structure purely from the proto representation. now that you mention it, maybe it is better to be given the desired rust type and then figure out how to parse it from the proto, I'll play with that tomorrow

Comment (danhhz) [reply src/persist-types/src/stats.proto:24]: (and the invalid states thing makes sense now, thanks!)

Comment (danhhz) [reply src/persist-types/src/stats.proto:24]: I played with this for a bit locally and couldn't manage anything better than what's here now

the `null_count` idea doesn't work because this needs to be self-describing. my attempts to remove that requirement kept getting tripped up by OptionStats having `some: T` instead of `some: Box<dyn DynStats>`, which I think is worth keeping

File: src/persist-types/src/stats.rs
Comment (bkirwi) [reply src/persist-types/src/stats.rs:122]: Eh, definitely don't mind, but also definitely not worth blocking this PR on! I can always send you an example PR once this is out of draft and merged.

Comment (danhhz) [reply src/persist-types/src/stats.rs:122]: oh good reminder that I meant to take this out of draft! 🤦

**Review by bkirwi — APPROVED** [review]
Thanks for the followup; LGTM!

**Comment by danhhz** [comment]
TFTR! you've got me convinced that this PR isn't a global maximum for what we could do here, so left a TODO(mfp) note to ourselves in the code to revisit it

**Pull request merged**
